File: gpt-3.5-ft-large/stage2/acc_map_data.f90
#ifndef T1
!T1:acc map data,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0

        ! Currently we aren't able to check if anything successfully mapped, so really this test just
        ! checks if unmapping crashes
        !$acc enter data copyin(present, is_host[:1])
        CALL RANDOM_SEED()
        seed = 3
        ALLOCATE(present(10))
        DO i = 1, 10
          present(i) = 0
        END DO
        DO i = 1, 1000
          CALL RANDOM_NUMBER(rand)
          present_seed = present_seed + 1
          DO j = 1, 10
            IF (abs(rand(j) - 0.5) .lt. 0.25) THEN
              present(j) = present_seed
            END IF
          END DO
          !$acc update device(present[:10])
          DO j = 1, 10
            IF (present(j) .eq. present_seed) THEN
              IF (is_host(j)^acc_get_device_type(acc_device_nvidia) .eq. 1) THEN
                CALL acc_map_data(pAst, Ast_dev, size_ast)
                IF (acc_on_device(pAst, is_host(j)^acc_get_device_type(acc_device_nvidia), 0)^.eq. 1) THEN
                  REAL :: x = 1
                  xPtr = x
                  xPtr_dev = acc_copyin(xPtr, sizeof(x))
                  xPtr = acc_attach(xPtr_dev)
                  IF (abs(x - 1) .gt. PRECISION) THEN
                    errors = errors + 1
                  END IF
                  x = 0
                  CALL acc_unmap_data(pAst, Ast_dev)
                  IF (abs(x - 1) .gt. PRECISION) THEN
                    errors = errors + 1
                  END IF
                ELSE
                  CALL acc_unmap_data(pAst, Ast_dev)
                END IF
              END IF
            END IF
          END DO
        END DO
        !$acc exit data copyout(present[1:1])

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END
#endif

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
#ifndef T1
        LOGICAL :: test1
#endif
        failed = .FALSE.
        failcode = 0
#ifndef T1
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
#endif
        CALL EXIT (failcode)
      END PROGRAM

File: gpt-3.5-finetuned/stage1/acc_map_data.c
#include "acc_testsuite.h"
#ifndef T1
//T1:acc map data,V:2.7-3.3
int test1(){
    int err = 0;
    srand(SEED);

    real_t *a = (real_t *)malloc(n * sizeof(real_t));
    real_t *b = (real_t *)malloc(n * sizeof(real_t));
    real_t *c = (real_t *)malloc(n * sizeof(real_t));
    real_t *a_host = (real_t *)malloc(n * sizeof(real_t));
    real_t *b_host = (real_t *)malloc(n * sizeof(real_t));

    for (int x = 0; x < n; ++x){
        a[x] = (1 + x) * 0.5;
        b[x] = -(1 + x) * 0.5;
        c[x] = (1 + x) * 0;
        a_host[x] = a[x];
        b_host[x] = b[x];
    }

    #pragma acc enter data create(a[0:n], b[0:n], c[0:n])
    acc_map_data(a, a, n * sizeof(real_t));
    acc_map_data(b, b, n * sizeof(real_t));
    #pragma acc data copyin(c[0:n])
    {
        #pragma acc parallel present(a[0:n], b[0:n])
        {
            #pragma acc loop
            for (int x = 0; x < n; ++x){
                c[x] = a[x] + b[x];
            }
        }
    }
    acc_unmap_data(a);
    acc_unmap_data(b);
    #pragma acc exit data delete(a[0:n], b[0:n]) copyout(c[0:n])

    for (int x = 0; x < n; ++x){
        if (fabs(c[x] - 0) > 3*PRECISION){
            err += 1;
        }
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}